### ISSUE TITLE
Collisions fail on perlmutter a100 GPUs

### ISSUE THREAD

**Issue by pordyna** [ComputationalRadiationPhysics/picongpu#4576]
## Description
While simulating a hydrogen jet target on perlmutter, the simulation will crash not at an exactly the same but more or less predictable moment (depends on the setup). The sim crashes in a collision kernel with:

```
[cupla] Error: </global/homes/p/pordyna/picongpu/include/picongpu/../pmacc/exec/KernelLauncher.tpp>:119 Crash after kernel launch N5pmacc8lockstep4exec6detail14LockStepKernelIN8picongpu9particles9collision14InterCollisionILb1EEENS0_9WorkerCfgILj256EEEEE [/global/homes/p/pordyna/picongpu/include/picongpu/../picongpu/particles/collision/InterCollision.hpp:489 ]
terminate called after throwing an instance of 'std::runtime_error'
  what():  /global/u2/p/pordyna/picongpu/thirdParty/cupla/alpaka/include/alpaka/event/EventUniformCudaHipRt.hpp(148) 'ret = TApi::eventQuery(event.getNativeHandle())' returned error  : 'cudaErrorIllegalAddress': 'an illegal memory access was encountered'!
```

## Setup
I was running with https://github.com/pordyna/picongpu/commit/93eb5ea5d51114288406c7b3b5e6eccaed5c39ef that is basically a very recent dev without any modifications except for the perlmutter profile.
The simulation setup was: [setup_for_git.zip](https://github.com/ComputationalRadiationPhysics/picongpu/files/11565631/setup_for_git.zip) and it was compiled with the `-t 0` cmake profile with the following output [compile_noDebug.log](https://github.com/ComputationalRadiationPhysics/picongpu/files/11565711/compile_noDebug.log)  ( the output above was produced with extra `PMACC_BLOCKING_KERNEL=ON, alpaka_CUDA_KEEP_FILES=ON, alpaka_CUDA_SHOW_CODELINES=ON`).
With `./bin/picongpu -d 1 1 1 -g 1440 1440 -s 10000 --periodic 0 0 -p 1` it crashes at around 20%.

My perlmutter setup:
[gpu.profile.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/11565717/gpu.profile.txt)
[pic_dependencies_autoinstall.sh.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/11565723/pic_dependencies_autoinstall.sh.txt)

## Debugging

Tried compiling with ` -DCMAKE_CUDA_FLAGS='-g -G -cudart shared' -DCMAKE_CXX_FLAGS='-g -O0' -DCMAKE_C_FLAGS='-g -O0' -DPIC_USE_ISAAC=OFF -DPIC_USE_PNGwriter=OFF -DPIC_USE_openPMD=OFF -DPMACC_BLOCKING_KERNEL=ON`, but it looks like
picongpu doesn't work with CUDA when compiling with debug info at all. I will open a separate issue for that and link it here.

## Other systems
The same simulation runs on hemera and the fwkt_v100 queue without any problem. It also runs on perlmutter on CPUs (with `omp2b backend`). I will also try it out on taurus, but I need to set up my picongpu there first.

**Comment by psychocoderHPC** [comment]
If you compile with debug you should increase the free memory https://github.com/ComputationalRadiationPhysics/picongpu/blob/267cb91a3f0c1dc00827a26e6e1242cfb11b816d/include/picongpu/param/memory.param#L43

The problem is compiling with DEBUG will produce many kernels that require local memory to spill registers, this memory is allocated by the driver on runtime. Due to the fact that PIConGPU is by default allocating the total GPU main memory except for 350MiB there is no memory left for dynamic allocations during the runtime.

This answer is only pointing to the second point of your issue where you compile without optimizations and should be unrelated to the first issue: `'an illegal memory access was encountered'!`

**Comment by pordyna** [comment]
@psychocoderHPC ok thanks, I will try increasing reserved memory

**Comment by psychocoderHPC** [comment]
A simple hack without recompiling is to add the parameter `-r 2` to PIConGPU which will tell PIConGPU to consume only half of the available GPU memory.

**Comment by PrometheusPi** [comment]
@pordyna Did reducing initial memory allocation helped?

**Comment by pordyna** [comment]
@PrometheusPi this was just related to why the version compiled with debug options didn't compile, this won't fix the actual collision problem I guess...

**Comment by psychocoderHPC** [comment]
I was able to reproduce the crash on our A30 GPU with the 2D setup where Intra collision was disabled

```bash
pic-build  -c "-DPMACC_BLOCKING_KERNEL=ON, -Dalpaka_CUDA_KEEP_FILES=ON, -Dalpaka_CUDA_SHOW_CODELINES=ON -DCMAKE_CXX_FLAGS=-g -DCMAKE_CUDA_FLAGS=-g"
./bin/picongpu -d 1 1 1 -g 512 720 -s 10000 --periodic 0 0 -p 1 -r 4
```

Crash after 14-15% with a possibility of 75%.


Next step:

```bash
# check only kernel with substring name InterCollision
compute-sanitizer --kernel-regex kns=InterCollision ./bin/picongpu -d 1 1 1 -g 512 720 -s 10000 --periodic 0 0 -p 1 -r 4
```

**Comment by psychocoderHPC** [comment]
I am currently using https://github.com/psychocoderHPC/picongpu/tree/fix-collisionMemAllocationIssue2 branch for debugging where I reduced the example and added memory pattern detection.

```
pic-build  -t 0 -c "-DPMACC_BLOCKING_KERNEL=ON, -Dalpaka_CUDA_KEEP_FILES=ON, -Dalpaka_CUDA_SHOW_CODELINES=ON -DCMAKE_CXX_FLAGS=-g -DCMAKE_CUDA_FLAGS=-g -DPIC_USE_openPMD=OFF" -f
for': x=0; for((i=0;i<100000&&x==0;i++)) ; do ./bin/picongpu -d 1 1 1 -g 720 720 -s 3000 --periodic 1 0 -p 1 -r 3 ; x=$?;echo $i; done 2>&1 | tee out.txt
```
The branch already contains a fix for mallocMC but to have documentation for my changes in collision I added the branch here.

**Comment by psychocoderHPC** [comment]
solved with #4634
